xdsl/utils/disjoint_set.py

```diff
@@ -0,0 +1,170 @@
+"""
+Generic implementation of a disjoint set data structure.
+
+https://en.wikipedia.org/wiki/Disjoint-set_data_structure
+"""
+
+from collections.abc import Hashable, Sequence
+from typing import Generic, TypeVar
+
+
+class IntDisjointSet:
```

Comment:
Is the IntDisjointSet independently useful? Otherwise this feels like an unnecessary step to making the structure below.

Reply:
Yeah, good point. I'm not sure why the generality here.
Also, if this is for the equality saturation tasks, I'm not familiar how it is used there; maybe future uses warrant this?

Reply:
I feel like it's a nice separation if only for testing purposes, it neatly encapsulates the core logic, and adds negligeable overhead. I expect that no-one's going to use it, but I also see no harm in the separation.

For equality saturation, my understanding is that a persistent version of this tends to be used, and I also expect that a separation between the core logic and the hashable/generic helpers would be useful both for testing and readability.